src/PluginClientBuilder.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpClient;
+use Http\Client\HttpAsyncClient;
+use Psr\Http\Client\ClientInterface;
+
+/**
+ * Build an instance of a PluginClient with a dynamic list of plugins.
+ *
+ * This builder is immutable, each "build" methods will instead create a new
+ * instance of itself with or without the affected plugin, destroying a
+ * potentially http client built in the new instance.
+ *
+ * @author Baptiste Clavié <clavie.b@gmail.com>
+ */
+final class PluginClientBuilder
+{
+    /** @var Plugin[] */
+    private $plugins = [];
+
+    /** @var ?PluginClient */
+    private $client;
+
+    public function addPlugin(Plugin $plugin): self
+    {
+        $builder = new self();
+
+        $builder->client = null;
+        $builder->plugins = $this->plugins;
+        $builder->plugins[get_class($plugin)] = $plugin;
+
+        return $builder;
+    }
+
+    public function removePlugin(string $fqcn): self
+    {
+        $builder = new self();
+
+        $builder->client = null;
+        $builder->plugins = $this->plugins;
+        unset($builder->plugins[$fqcn]);
+
+        return $builder;
+    }
+
+    /**
+     * @param ClientInterface | HttpClient | HttpAsyncClient $client Client to nest into the plugin client
+     */
+    public function createClient($client, array $options = []): PluginClient
+    {
+        if (!$client instanceof ClientInterface && !$client instanceof HttpClient && !$client instanceof HttpAsyncClient) {
+            throw new \RuntimeException('You must provide a valid http client');
+        }
+
+        if (null === $this->client) {
```

Comment:
The builder is immutable at least it tries to be when it comes to modifying the plugin chain, but it isn't when you create a client. Also, if I want to use the builder multiple times and I pass in different client instances I won't get a plugin client with that specific client, but one with an eariler client instance without a warning. This kind of behavior easily leads to confusion IMHO.

Reply:
This is to avoid rebuilding the client each time. To solve this, I could add a method to set the http client on the builder itself (while making it immutable), and use that to build the plugin client (with a cache so it doesn't get build on each calls)